src/main/java/ch/njol/skript/lang/FunctionArgumentParser.java

```diff
@@ -0,0 +1,160 @@
+package ch.njol.skript.lang;
+
+import org.skriptlang.skript.lang.function.FunctionReference.Argument;
+import org.skriptlang.skript.lang.function.FunctionReference.ArgumentType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Parses the arguments of a function reference.
+ */
+final class FunctionArgumentParser {
+
+	/**
+	 * The input string.
+	 */
+	private final String args;
+
+	/**
+	 * The list of unmapped arguments.
+	 */
+	private final List<Argument<String>> arguments = new ArrayList<>();
+
+	/**
+	 * The char index.
+	 */
+	private int index = 0;
+
+	/**
+	 * Constructs a new function argument parser based on the
+	 * input string and instantly calculates the result.
+	 *
+	 * @param args The input string.
+	 */
+	public FunctionArgumentParser(String args) {
+		this.args = args;
+
+		parse();
+	}
+
+	private char c;
+	private boolean nameFound = false;
+	private final StringBuilder namePart = new StringBuilder();
+	private final StringBuilder exprPart = new StringBuilder();
+
+	private boolean inString = false;
+	private int nesting = 0;
+
+	private void parse() {
```

Comment:
these functions need some comprehensive docs to describe their behavior and goals